lib/puppet/context.rb

```diff
@@ -72,12 +72,14 @@ def mark(name)
   #
   # @api private
   def rollback(name)
-    if @rollbacks[name].nil?
+    if @rollbacks.value[name].nil?
       raise UnknownRollbackMarkError, _("Unknown mark '%{name}'") % { name: name }
     end
 
-    @stack = @rollbacks[name]
-    @rollbacks.delete(name)
+    @stack.value = @rollbacks.value[name]
+    rollbacks_without_name = @rollbacks.value.dup
+    rollbacks_without_name.delete(name)
+    @rollbacks.value = rollbacks_without_name
```

Comment:
Would `@rollbacks.value.delete(name)` work?

Reply:
no, delete mutates the hash which causes this to leak between threads

Reply:
Should `rollbacks` be dup'ed in the `mark` method too? Otherwise (I think?) if one thread calls `mark`, then that `name` will be visible to another thread? Would it make sense to freeze the hash before calling `@rollbacks.value = <hash>` so we don't accidentally mutate it?

Reply:
Yes, good catch

Reply:
I think the way I've done the rollbacks is totally wrong. Since you shouldn't roll back across threads I think I can just lazily initialize a new hash in each thread and then I don't have to worry about duplicating anything. Also this method is no longer used outside of our test setup.